### PR TITLE
clisqlshell: hide the libedit dep behind a go interface

### PR DESCRIPTION
**Pull request by knz**
In #86457 and related work we will want to offer two editors side-by-side in a transition period, so folk can compare or fall back on something known in case they are not happy with the new stuff.

To enable this transition period, this commit hides the editor behind a go interface.

This also makes the shell code overall easier to read and understand.

Epic: CRDB-22182
Release note: None

### CODE REVIEW

**Review by ZhouXing19 — APPROVED** [review]
Very cool improvement! It makes the code much easier to read.

File: pkg/cli/clisqlshell/editor_bufio.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/cockroachdb/errors"
+)
+
+// bufioReader implements the editor interface.
```
Comment (rafiss) [inline pkg/cli/clisqlshell/editor_bufio.go:22]: could we include declarations for the implementing types that assert that they implement the interface?

Comment (knz) [reply pkg/cli/clisqlshell/editor_bufio.go:22]: Good idea. Done.

**Comment by knz** [comment]
bors r=ZhouXing19,rafiss

**Pull request merged**